scripts/getDivviIntegrators.ts

```diff
@@ -0,0 +1,168 @@
+function removeDuplicates<T>(arr: T[]): T[] {
+  const seen = new Set<T>()
+  return arr.filter((item) => {
+    if (seen.has(item)) {
+      return false
+    }
+    seen.add(item)
+    return true
+  })
+}
+
+async function getDivviIntegrators({
+  rewardPoolContractAddress,
+}: {
+  rewardPoolContractAddress: string
+}): Promise<Address[]> {
+  const usersThatHaveIntegrated: Address[] = []
+  const usersThatHaveReceivedRewards = new Set<Address>()
+  const usersThatHaveRegisteredAgreements = new Set<Address>()
+
+  const REFERRAL_REGISTERED_TOPIC = encodeEventTopics({
+    abi: divviRegistryAbi,
+    eventName: 'ReferralRegistered',
+  })[0]
+
+  const queryForIntegrators = {
+    logs: [
+      {
+        address: [DIVVI_REGISTRY_CONTRACT_ADDRESS],
+        topics: [[REFERRAL_REGISTERED_TOPIC]],
+      },
+    ],
+    fieldSelection: {
+      log: [LogField.Topic0, LogField.Topic1, LogField.Topic2, LogField.Topic3],
+    },
+    fromBlock: 0,
+  }
+
+  const ADD_REWARD_TOPIC = encodeEventTopics({
+    abi: rewardPoolAbi,
+    eventName: 'AddReward',
+  })[0]
+
+  const queryForRewardsReceivers = {
+    logs: [
+      { address: [rewardPoolContractAddress], topics: [[ADD_REWARD_TOPIC]] },
+    ],
+    fieldSelection: {
+      log: [LogField.Topic0, LogField.Topic1],
+    },
+    fromBlock: 0,
+  }
+
+  const REWARDS_AGREEMENT_REGISTERES_TOPIC = encodeEventTopics({
+    abi: divviRegistryAbi,
+    eventName: 'RewardsAgreementRegistered',
+  })[0]
+
+  const queryForRegisteredAgreements = {
+    logs: [
+      {
+        address: [DIVVI_REGISTRY_CONTRACT_ADDRESS],
+        topics: [
+          [REWARDS_AGREEMENT_REGISTERES_TOPIC],
+          [DIVVI_INTEGRATION_REWARDS_ENTITY],
+        ],
+      },
+    ],
+    fieldSelection: {
+      log: [LogField.Topic0, LogField.Topic1, LogField.Topic2],
+    },
+    fromBlock: 0,
+  }
+
+  const client = getHyperSyncClient(NetworkId['op-mainnet'])
+
+  await Promise.all([
+    paginateQuery(client, queryForIntegrators, async (response) => {
+      for (const transaction of response.data.logs) {
+        usersThatHaveIntegrated.push(transaction.topics[3] as Address)
+      }
+    }),
+    paginateQuery(client, queryForRewardsReceivers, async (response) => {
+      for (const transaction of response.data.logs) {
+        usersThatHaveReceivedRewards.add(transaction.topics[1] as Address)
+      }
+    }),
+    paginateQuery(client, queryForRegisteredAgreements, async (response) => {
+      for (const transaction of response.data.logs) {
+        usersThatHaveRegisteredAgreements.add(transaction.topics[2] as Address)
+      }
+    }),
+  ])
+
+  const deduplicatedUsersThatHaveIntegrated = removeDuplicates(
+    usersThatHaveIntegrated,
+  )
+
+  // TODO(ENG-345): Also filter for if the user is whitelisted
+  const userToReceiveRewards = deduplicatedUsersThatHaveIntegrated.filter(
+    (user: Address) =>
+      !usersThatHaveReceivedRewards.has(user) &&
+      usersThatHaveRegisteredAgreements.has(user),
+  )
```

Comment:
do we need to track the registered agreements? the contract already ensures that [there is an agreement](https://github.com/divvi-xyz/divvi-protocol-v0/blob/57649c63778eccbfa05c0b4dcfd798e89ae5ba24/contracts/DivviRegistry.sol#L268-L273) before registering the referral

Reply:
ooohhhh i see, `usersThatHaveRegisteredAgreements` is to track which entities have agreements specifically with the divvi provider. can we rename the variable to be more specific? e.g. `divviAgreementConsumers` / `consumersWithDivviAgreements`?